File: src/upros_tutorial/src/ros_action_server.cpp
#include <ros/ros.h>
#include <actionlib/server/simple_action_server.h>
#include <upros_message/MyActionAction.h>

class MyActionServer
{
protected:
    ros::NodeHandle nh_;
    actionlib::SimpleActionServer<upros_message::MyActionAction> as_;
    std::string action_name_;
    upros_message::MyActionFeedback feedback_;
    upros_message::MyActionResult result_;

public:
    MyActionServer(std::string name) : as_(nh_, name, boost::bind(&MyActionServer::executeCB, this, _1), false),
                                       action_name_(name)
    {
        as_.start();
    }

    ~MyActionServer(void) {}

    void executeCB(const upros_message::MyActionGoalConstPtr &goal)
    {
        ros::Rate r(1);
        bool success = true;

        // 执行动作
        for (int i = 1; i <= 10; i++)
        {
            if (as_.isPreemptRequested() || !ros::ok())
            {
                ROS_INFO("%s: Preempted", action_name_.c_str());
                as_.setPreempted();
                success = false;
                break;
            }

            feedback_.progress = i * 10;
            ROS_INFO("%s: Executing, progress = %f", action_name_.c_str(), feedback_.progress);
            as_.publishFeedback(feedback_);

            r.sleep();
        }

        // 发送结果
        if (success)
        {
            result_.success = true;
            ROS_INFO("%s: Succeeded", action_name_.c_str());
            as_.setSucceeded(result_);
        }
    }
};

int main(int argc, char **argv)
{
    ros::init(argc, argv, "my_action_server");
    MyActionServer server("my_action");
    ros::spin();
    return 0;
}

File: src/upros_tutorial/scripts/ros_action_server.py
#!/usr/bin/env python3

import rospy
import actionlib

from upros_message.msg import MyActionAction,MyActionGoal,MyActionResult,MyActionFeedback

class MyActionServer(object):
    def __init__(self, name):
        self._action_name = name
        self._as = actionlib.SimpleActionServer(self._action_name, MyActionAction, execute_cb=self.executeCB, auto_start = False)
        self._as.start()
        self._feedback = MyActionFeedback()
        self._result = MyActionResult()

    def executeCB(self, goal):
        r = rospy.Rate(1)
        success = True
        # 执行动作
        for i in range(1, 11):
            if self._as.is_preempt_requested():
                rospy.loginfo('{}: Preempted'.format(self._action_name))
                self._as.set_preempted()
                success = False
                break

            self._feedback.progress = i * 10
            rospy.loginfo('{}: Executing, progress = {}'.format(self._action_name, self._feedback.progress))
            self._as.publish_feedback(self._feedback)

            r.sleep()

        # 发送结果
        if success:
            self._result.success = True
            rospy.loginfo('{}: Succeeded'.format(self._action_name))
            self._as.set_succeeded(self._result)

if __name__ == '__main__':
    rospy.init_node('my_server')
    server = MyActionServer('my_action')
    rospy.spin()
